### PR TITLE
XSPI (OctoSPI/Quad SPI/Dual SPI) support

### PR DESCRIPTION
**Pull request by NagaHimanshu**
## Pull Request Description

- Added no_os_xspi to drivers/api
- Added support for xspi on stm32 platform
- Tested on Nucleo-H563ZI board

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: include/no_os_xspi.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
+/***************************************************************************//**
```
Comment (buha) [inline include/no_os_xspi.h:1]: I don't think we should add no_os_xpi.h/.c

We can just as well use no_os_spi.h module for the same purpose.
I've diffed your xspi.h to spi.h and it seems the only fundamental change is that you added a specifier for the number of lanes (single, bi, quad, octo). This could be easily done by adding a `lanes` member to the spi init_param and desc, which by default/uninitialized would mean single lane to be backwards compatible.

Then you'd proceed to implement your own implementation of the no_os_spi API in stm32_octospi for example, and populate the platform ops. Keep in mind that the same no_os_spi API could be implemented as xilinx_octalspi, so no_os_spi needs to be as generic as possible.

File: drivers/platform/stm32/stm32_xspi.c [hunk 2]
```diff
@@ -0,0 +1,718 @@
+/***************************************************************************//**
```
Comment (buha) [inline drivers/platform/stm32/stm32_xspi.c:1]: so this is good in principle, I'd rename it and the function ops to stm32_octospi to reflect the true peripheral being used

and as I said in the other comment, make it use generic (and slightly modified) no_os_spi instead of a new no_os_xspi

File: include/no_os_xspi.h (hunk 1 above)
Comment (NagaHimanshu) [reply include/no_os_xspi.h:1]: Sure, this is a great idea and can be done. 

In STM HAL files, they have referred it to as XSPI driver and I was using the same convention as _xspi in this case. Will it be better to use the same or shall we proceed with _octospi?

Comment (buha) [reply include/no_os_xspi.h:1]: Maybe STM has very similar implementations for dual/quad/octo SPI's with very few register map differences and behavior, in which case it's better to make one stm32_xspi implementation that handles all of them, yes. But if they have wildly different implementations then it's better to keep them separate. I'm not very familiar with their offer.

File: drivers/platform/stm32/stm32_xspi.c [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+	timeout = msgs->bytes_number;
+	while(timeout--) {
+		no_os_mdelay(1);
+		if (sdesc->stm32_xspi_dma_done)
+			break;
+	};
+
+	/* need some cleanup here? */
+	if (timeout == 0)
+		return -ETIME;
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.c:702]: You should also call `no_os_dma_abort()` if you reach this point.

File: drivers/platform/stm32/stm32_xspi.h [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+#include <stdint.h>
+#include <stdbool.h>
+#include "no_os_xspi.h"
+#include "stm32_hal.h"
+
+/**
+ * @struct stm32_xspi_command
+ * @brief Structure holding the stm32 platform specific XSPI command parameters.
+ */
+typedef struct stm32_xspi_command {
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.h:51]: We don't allow `typedef`. Just use `struct stm32_xspi_command` when declaring a variable of this type (as you're already doing). Moreover, it doesn't look like you want to expose this type to the user. In this case, please move it to the source file.

File: drivers/platform/stm32/stm32_xspi.h [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	/** Get peripheral source clock function */
+	uint32_t (*get_input_clock)(void);
+	/** FIFO threshold */
+	uint32_t fifo_threshold;
+	/** Sample Shifting */
+	bool sample_shifting;
+	/** CS high time (in number of clock cycles) */
+	uint32_t cs_high_time;
+	/** DMA Initialization Parameters */
+	struct no_os_dma_init_param* dma_init;
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.h:95]: `struct no_os_dma_init_param *dma_init;`

File: drivers/platform/stm32/stm32_xspi.c [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+			sdesc->dma_ch->irq_num = sinit->irq_num;
+		}
+	}
+
+	*desc = xspi_desc;
+
+	return 0;
+
+error:
+	no_os_dma_remove(sdesc->dma_desc);
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.c:323]: `sdesc` may not be allocated at this point if `xspi_desc` alloc fails. Also, please initialize all the pointers you're calling `free()` on here to `NULL`. It's undefined behavior to free a random pointer (which doesn't point to a memory region previously allocated).

File: drivers/platform/stm32/stm32_xspi.c [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+	return 0;
+}
+
+/**
+ * @brief Configure XSPI with new parameters.
+ * @param desc - The XSPI descriptor
+ * @param params - The structure that contains the XSPI parameters.
+ * @return 0 in case of success, negative code otherwise.
+ */
+int32_t stm32_xspi_configure(struct no_os_xspi_desc *desc,
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.c:152]: All the functions in this file should be `static`

File: include/no_os_spi.h [hunk 8]
```diff
@@ -84,6 +84,21 @@ enum no_os_spi_bit_order {
 	NO_OS_SPI_BIT_ORDER_LSB_FIRST = 1,
 };
 
+/**
+ * @enum no_os_spi_lanes
+ * @brief SPI configuration for number of lanes.
+ */
+enum no_os_spi_lanes {
+	/** Single Lane */
+	NO_OS_SPI_SINGLE_LANE = 0,
```
Comment (buha) [inline include/no_os_spi.h:93]: no need to specify their values

File: include/no_os_spi.h [hunk 9]
```diff
@@ -187,7 +208,11 @@ struct no_os_spi_desc {
 	enum no_os_spi_mode	mode;
 	/** SPI bit order */
 	enum no_os_spi_bit_order	bit_order;
+	/** SPI Lanes */
+	enum no_os_spi_lanes   lanes;
+	/** SPI bus platform ops */
```
Comment (buha) [inline include/no_os_spi.h:213]: these documentation strings should go in a separate commit, something like "document all no_os_spi structure members"

File: drivers/platform/stm32/stm32_xspi.c [hunk 10]
```diff
@@ -5,36 +5,30 @@
 ********************************************************************************
  * Copyright 2024(c) Analog Devices, Inc.
  *
- * All rights reserved.
```
Comment (buha) [inline drivers/platform/stm32/stm32_xspi.c]: please squash this commit into the commit that adds stm32_xspi implementation

since you add that implementation, add it with the proper license (3 clause BSD) from the start

**Review by buha — DISMISSED** [review]

File: include/no_os_spi.h [hunk 11]
```diff
@@ -233,6 +253,10 @@ int32_t no_os_spi_init(struct no_os_spi_desc **desc,
 /* Free the resources allocated by no_os_spi_init(). */
 int32_t no_os_spi_remove(struct no_os_spi_desc *desc);
 
+/* Configure the SPI peripheral */
+int32_t no_os_spi_config(struct no_os_spi_desc *desc,
+			 const struct no_os_spi_init_param *param);
+
```
Comment (CiprianRegus) [inline include/no_os_spi.h:266]: Can you add more details about what this new function does and why is it needed? "Configure the SPI peripheral" is too vague and it doesn't look like you can actually reconfigure the fields defined in `struct no_os_spi_desc` (only those for xspi).

**Review by CiprianRegus — CHANGES_REQUESTED** [review]

File: drivers/platform/stm32/stm32_xspi.c [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+	struct stm32_xspi_desc *sdesc;
+	XSPI_RegularCmdTypeDef *cmd;
+
+	if (!desc || !desc->extra || !msgs)
+		return -EINVAL;
+
+	sdesc = desc->extra;
+	cmd = sdesc->command;
+
+	for (uint32_t i = 0; i < len; i++) {
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_xspi.c:380]: For the standard SPI peripheral we're configuring the hardware registers before the transfer (based on the fields in the no os and STM32 specific SPI descriptors). I think we could do the same in this case and avoid the need to have the new `configure` API.

Comment (NagaHimanshu) [reply drivers/platform/stm32/stm32_xspi.c:380]: The original intent for adding configure API is to reconfigure the vendor specific params without reinitializing the peripheral and creating descriptors again. If this is the case, instead of `no_os_spi_init_params` as argument, is it good to have `void *params` just to reconfigure vendor specific params?

If configuration need to be done, then user will use `no_os_spi_init` with new set of init params.

Comment (buha) [reply drivers/platform/stm32/stm32_xspi.c:380]: I think you could do what @CiprianRegus says for now and drop it. A configure function would be very useful but not necessarily how it's implemented. The way I see it is that for each peripheral driver we could have:
- init() - allocate memory, calls config() internally
- config() - configure the peripheral
- remove() - deallocate memory (and put the hardware in a good state, if needed)

The user could then reconfigure the peripheral with a new set of init_param structure without deallocating/reallocating memory. 

But this is quite a big change and I don't think this PR should do such big changes, just drop the config function for now and we'll see about it in the future.

Comment (buha) [reply drivers/platform/stm32/stm32_xspi.c:380]: PS: the idea was good, we even have an issue with the UART where we'd like to reconfigure the baudrate without reallocating it, like you proposed for this spi, it's just that we need to have a proper discussion about this internally

**Review by buha — APPROVED** [review]
Ciprian's request for changes was made, LGTM so I'm merging

**Review by CiprianRegus — APPROVED** [review]
